File: sdk/python/clampd/stream_guard.py
"""Streaming tool call interception for OpenAI and Anthropic."""

from __future__ import annotations

import json
import logging
from collections.abc import Iterator
from typing import Any

from clampd.client import ClampdBlockedError, ClampdClient

logger = logging.getLogger("clampd")


class _StreamProxy:
    """Wraps a stream iterator and proxies attributes from the original object."""

    def __init__(self, original: Any, generator: Iterator[Any]) -> None:
        self._original = original
        self._generator = generator

    def __iter__(self) -> Iterator[Any]:
        return self._generator

    def __next__(self) -> Any:
        return next(self._generator)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._original, name)

    def __enter__(self) -> _StreamProxy:
        if hasattr(self._original, "__enter__"):
            self._original.__enter__()
        return self

    def __exit__(self, *exc: object) -> None:
        if hasattr(self._original, "__exit__"):
            self._original.__exit__(*exc)


def guard_openai_stream(
    stream: Any,
    client: ClampdClient,
    *,
    agent_id: str = "",
    target_url: str = "",
    fail_open: bool = False,
    authorized_tools: list[str] | None = None,
) -> Any:
    """Wraps an OpenAI streaming response to intercept tool calls.

    Text chunks pass through immediately. Tool call chunks are buffered
    until the stream signals finish_reason="tool_calls", at which point
    each tool call is sent through client.proxy() for policy evaluation.
    If blocked, raises ClampdBlockedError. If allowed, yields the
    buffered tool call chunks.

    Uses a SHARED buffer (not per-tool-call) for two reasons:

    1. OpenAI sends a single chunk with deltas for multiple tool calls.
       Per-tool-call buffers would either lose chunks (only store in one)
       or duplicate them (store in all, yield the same chunk N times).

    2. Parallel tool calls in the same LLM response are atomic — if the
       model says "call weather AND calendar", both must be approved before
       either executes. If one is denied, the entire response is blocked.
       Partial release of one tool call's chunks would give the consumer
       an incomplete response that can't be acted on.
    """

    def _generate() -> Iterator[Any]:
        # Accumulate tool call data per index; shared buffer holds all chunks.
        tool_calls: dict[int, dict[str, Any]] = {}
        buffered_chunks: list[Any] = []  # shared — see docstring for why

        for chunk in stream:
            choice = chunk.choices[0] if chunk.choices else None
            if choice is None:
                yield chunk
                continue

            delta = choice.delta if hasattr(choice, "delta") else None
            if delta is None:
                yield chunk
                continue

            # Check if this chunk has tool call deltas
            tc_deltas = getattr(delta, "tool_calls", None)
            if tc_deltas:
                buffered_chunks.append(chunk)
                for tc in tc_deltas:
                    idx = tc.index
                    if idx not in tool_calls:
                        tool_calls[idx] = {"name": "", "arguments": ""}
                    fn = getattr(tc, "function", None)
                    if fn:
                        if getattr(fn, "name", None):
                            tool_calls[idx]["name"] = fn.name
                        if getattr(fn, "arguments", None):
                            tool_calls[idx]["arguments"] += fn.arguments

                # Check if stream is done with tool calls
                if getattr(choice, "finish_reason", None) == "tool_calls":
                    # Guard each accumulated tool call
                    for idx in sorted(tool_calls):
                        tc_data = tool_calls[idx]
                        tool_name = tc_data["name"]
                        try:
                            tool_args = json.loads(tc_data["arguments"]) if tc_data["arguments"] else {}
                        except (json.JSONDecodeError, TypeError):
                            tool_args = {"_raw": tc_data["arguments"]}

                        try:
                            result = client.proxy(
                                tool=tool_name,
                                params=tool_args,
                                target_url=target_url,
                                authorized_tools=authorized_tools,
                            )
                        except ClampdBlockedError:
                            raise
                        except Exception as e:
                            if fail_open:
                                logger.warning("Clampd gateway error (fail-open): %s", e)
                                continue
                            raise ClampdBlockedError(str(e)) from e

                        if not result.allowed:
                            raise ClampdBlockedError(
                                result.denial_reason or "denied",
                                risk_score=result.risk_score,
                                response=result,
                            )

                    # All tool calls allowed — yield buffered chunks
                    yield from buffered_chunks
                    buffered_chunks.clear()
                    tool_calls.clear()
            else:
                # Text or non-tool chunk — pass through immediately
                yield chunk

    return _StreamProxy(stream, _generate())


def guard_anthropic_stream(
    stream: Any,
    client: ClampdClient,
    *,
    agent_id: str = "",
    target_url: str = "",
    fail_open: bool = False,
    authorized_tools: list[str] | None = None,
) -> Any:
    """Wraps an Anthropic streaming response to intercept tool_use blocks.

    Text events pass through immediately. Tool use events are buffered
    until content_block_stop, at which point the tool call is sent through
    client.proxy() for policy evaluation. If blocked, raises ClampdBlockedError.
    If allowed, yields the buffered events.
    """

    def _generate() -> Iterator[Any]:
        # Current tool block being accumulated
        current_tool_name: str = ""
        current_tool_json: str = ""
        buffered_events: list[Any] = []
        in_tool_block: bool = False

        for event in stream:
            event_type = getattr(event, "type", None)

            if event_type == "content_block_start":
                block = getattr(event, "content_block", None)
                if block and getattr(block, "type", None) == "tool_use":
                    # Start buffering a tool_use block
                    in_tool_block = True
                    current_tool_name = getattr(block, "name", "")
                    current_tool_json = ""
                    buffered_events = [event]
                    continue

            if in_tool_block:
                buffered_events.append(event)

                if event_type == "content_block_delta":
                    delta = getattr(event, "delta", None)
                    if delta and getattr(delta, "type", None) == "input_json_delta":
                        current_tool_json += getattr(delta, "partial_json", "")

                elif event_type == "content_block_stop":
                    # Tool block complete — guard it
                    try:
                        tool_args = json.loads(current_tool_json) if current_tool_json else {}
                    except (json.JSONDecodeError, TypeError):
                        tool_args = {"_raw": current_tool_json}

                    try:
                        result = client.proxy(
                            tool=current_tool_name,
                            params=tool_args,
                            target_url=target_url,
                            authorized_tools=authorized_tools,
                        )
                    except ClampdBlockedError:
                        raise
                    except Exception as e:
                        if fail_open:
                            logger.warning("Clampd gateway error (fail-open): %s", e)
                            in_tool_block = False
                            continue
                        raise ClampdBlockedError(str(e)) from e

                    if not result.allowed:
                        raise ClampdBlockedError(
                            result.denial_reason or "denied",
                            risk_score=result.risk_score,
                            response=result,
                        )

                    # Allowed — yield buffered events
                    yield from buffered_events
                    in_tool_block = False
                    buffered_events = []
                    current_tool_name = ""
                    current_tool_json = ""

                continue

            # Non-tool event — pass through immediately
            yield event

    return _StreamProxy(stream, _generate())

File: sdk/typescript/src/stream-guard.ts
/**
 * Stream interception for OpenAI and Anthropic streaming responses.
 *
 * Accumulates tool call chunks from streaming responses, guards them
 * through the Clampd proxy, and only releases chunks once approved.
 * Text/content chunks pass through immediately with zero added latency.
 */

import { ClampdClient, type ProxyResponse } from "./client.js";
import { ClampdBlockedError } from "./interceptor.js";
import { withDelegation, getDelegation, getCallerAgentId } from "./delegation.js";

// ── Types ────────────────────────────────────────────────────────

export interface StreamGuardOptions {
  agentId: string;
  targetUrl?: string;
  failOpen?: boolean;
  authorizedTools?: string[];
}

/** Minimal OpenAI streaming chunk shape. */
interface OpenAIChunk {
  choices?: Array<{
    index?: number;
    delta?: {
      content?: string | null;
      tool_calls?: Array<{
        index: number;
        id?: string;
        type?: string;
        function?: {
          name?: string;
          arguments?: string;
        };
      }>;
    };
    finish_reason?: string | null;
  }>;
  [key: string]: unknown;
}

/** Accumulated tool call state for OpenAI. */
interface PendingOpenAIToolCall {
  index: number;
  name: string;
  argumentFragments: string[];
}

/** Minimal Anthropic streaming event shape. */
interface AnthropicStreamEvent {
  type: string;
  index?: number;
  content_block?: {
    type: string;
    id?: string;
    name?: string;
    input?: Record<string, unknown>;
  };
  delta?: {
    type?: string;
    text?: string;
    partial_json?: string;
    stop_reason?: string;
  };
  [key: string]: unknown;
}

/** Accumulated tool call state for Anthropic. */
interface PendingAnthropicToolCall {
  blockIndex: number;
  name: string;
  jsonFragments: string[];
  bufferedEvents: AnthropicStreamEvent[];
}

// ── Proxy helper ─────────────────────────────────────────────────

async function guardToolCall(
  client: ClampdClient,
  toolName: string,
  toolArgs: Record<string, unknown>,
  opts: StreamGuardOptions,
): Promise<void> {
  await withDelegation(opts.agentId, async () => {
    const delegation = getDelegation();
    const proxyParams: Record<string, unknown> = { ...toolArgs };
    if (delegation && delegation.chain.length > 1) {
      proxyParams._delegation = {
        caller_agent_id: getCallerAgentId(),
        delegation_chain: delegation.chain,
        delegation_trace_id: delegation.traceId,
      };
    }

    try {
      const res = await client.proxy(
        toolName,
        proxyParams,
        opts.targetUrl ?? "",
        undefined,
        undefined,
        opts.authorizedTools,
      );
      if (!res.allowed) {
        // Respect failOpen for gateway errors
        if (opts.failOpen && res._gatewayError) {
          return;
        }
        throw new ClampdBlockedError(res);
      }
    } catch (e) {
      if (e instanceof ClampdBlockedError) throw e;
      if (!opts.failOpen) throw new ClampdBlockedError({
        request_id: "error",
        allowed: false,
        risk_score: 1.0,
        denial_reason: String(e),
        latency_ms: 0,
        degraded_stages: [],
        session_flags: [],
      });
    }
  });
}

// ── OpenAI stream guard ──────────────────────────────────────────

/**
 * Wraps an OpenAI streaming async iterable to intercept tool calls.
 * Text chunks pass through immediately. Tool call chunks are buffered
 * until complete, then guarded through the proxy before being released.
 *
 * Uses a SHARED buffer (not per-tool-call) for two reasons:
 *
 * 1. OpenAI sends a single chunk with deltas for multiple tool calls.
 *    Per-tool-call buffers would either lose chunks (only store in one)
 *    or duplicate them (store in all, yield the same chunk N times).
 *
 * 2. Parallel tool calls in the same LLM response are atomic — if the
 *    model says "call weather AND calendar", both must be approved before
 *    either executes. If one is denied, the entire response is blocked.
 *    Partial release of one tool call's chunks would give the consumer
 *    an incomplete response that can't be acted on.
 */
export function guardOpenAIStream(
  stream: AsyncIterable<unknown>,
  client: ClampdClient,
  opts: StreamGuardOptions,
): AsyncIterable<unknown> {
  const guardedIterator = async function* (): AsyncIterableIterator<unknown> {
    const pending = new Map<number, PendingOpenAIToolCall>();
    // Shared buffer: all tool call chunks held until ALL tool calls are guarded.
    const bufferedChunks: OpenAIChunk[] = [];
    let hasToolCalls = false;

    for await (const rawChunk of stream) {
      const chunk = rawChunk as OpenAIChunk;
      const choice = chunk.choices?.[0];

      if (!choice?.delta?.tool_calls?.length) {
        // No tool call data — yield text/other chunks immediately
        // But only if we haven't started accumulating tool calls,
        // or if this is a non-tool chunk interleaved
        if (!hasToolCalls) {
          yield chunk;
          continue;
        }
        // Buffer finish chunks until tool calls are guarded
        if (choice?.finish_reason) {
          // Guard all accumulated tool calls now
          for (const [, tc] of pending) {
            const argsStr = tc.argumentFragments.join("");
            let toolArgs: Record<string, unknown>;
            try {
              toolArgs = JSON.parse(argsStr);
            } catch {
              toolArgs = { raw: argsStr };
            }
            await guardToolCall(client, tc.name, toolArgs, opts);
          }

          // Release all buffered chunks
          for (const buffered of bufferedChunks) {
            yield buffered;
          }
          bufferedChunks.length = 0;
          pending.clear();
          hasToolCalls = false;

          // Now yield the finish chunk
          yield chunk;
          continue;
        }

        yield chunk;
        continue;
      }

      // Tool call deltas present — buffer them
      hasToolCalls = true;
      for (const tcDelta of choice.delta.tool_calls) {
        const idx = tcDelta.index;
        if (!pending.has(idx)) {
          pending.set(idx, {
            index: idx,
            name: tcDelta.function?.name ?? "unknown",
            argumentFragments: [],
          });
        }
        const tc = pending.get(idx)!;
        if (tcDelta.function?.name && !tc.name) {
          tc.name = tcDelta.function.name;
        }
        if (tcDelta.function?.arguments) {
          tc.argumentFragments.push(tcDelta.function.arguments);
        }
      }
      // Buffer the entire chunk (released after guard approval)
      bufferedChunks.push(chunk);
    }

    // Stream ended — guard any remaining tool calls
    if (pending.size > 0) {
      for (const [, tc] of pending) {
        const argsStr = tc.argumentFragments.join("");
        let toolArgs: Record<string, unknown>;
        try {
          toolArgs = JSON.parse(argsStr);
        } catch {
          toolArgs = { raw: argsStr };
        }
        await guardToolCall(client, tc.name, toolArgs, opts);
      }
      for (const buffered of bufferedChunks) {
        yield buffered;
      }
    }
  };

  // Return a Proxy that intercepts Symbol.asyncIterator but delegates
  // everything else (e.g. .controller, .toReadableStream()) to the original
  return new Proxy(stream as object, {
    get(target, prop, receiver) {
      if (prop === Symbol.asyncIterator) {
        return () => guardedIterator();
      }
      return Reflect.get(target, prop, receiver);
    },
  }) as AsyncIterable<unknown>;
}

// ── Anthropic stream guard ───────────────────────────────────────

/**
 * Wraps an Anthropic streaming async iterable to intercept tool_use blocks.
 * Text events pass through immediately. Tool use events are buffered per-block,
 * guarded at content_block_stop, then released.
 */
export function guardAnthropicStream(
  stream: AsyncIterable<unknown>,
  client: ClampdClient,
  opts: StreamGuardOptions,
): AsyncIterable<unknown> {
  const guardedIterator = async function* (): AsyncIterableIterator<unknown> {
    let currentToolCall: PendingAnthropicToolCall | null = null;

    for await (const rawEvent of stream) {
      const event = rawEvent as AnthropicStreamEvent;

      switch (event.type) {
        case "content_block_start": {
          if (event.content_block?.type === "tool_use") {
            // Start buffering a tool_use block
            currentToolCall = {
              blockIndex: event.index ?? 0,
              name: event.content_block.name ?? "unknown",
              jsonFragments: [],
              bufferedEvents: [event],
            };
          } else {
            // Text block start — pass through
            yield event;
          }
          break;
        }

        case "content_block_delta": {
          if (currentToolCall && event.delta?.type === "input_json_delta") {
            // Accumulate JSON fragments for the tool call
            if (event.delta.partial_json) {
              currentToolCall.jsonFragments.push(event.delta.partial_json);
            }
            currentToolCall.bufferedEvents.push(event);
          } else {
            // Text delta — pass through
            yield event;
          }
          break;
        }

        case "content_block_stop": {
          if (currentToolCall && event.index === currentToolCall.blockIndex) {
            // Tool block complete — guard it
            currentToolCall.bufferedEvents.push(event);
            const argsStr = currentToolCall.jsonFragments.join("");
            let toolArgs: Record<string, unknown>;
            try {
              toolArgs = argsStr ? JSON.parse(argsStr) : {};
            } catch {
              toolArgs = { raw: argsStr };
            }

            await guardToolCall(client, currentToolCall.name, toolArgs, opts);

            // Allowed — release buffered events
            for (const buffered of currentToolCall.bufferedEvents) {
              yield buffered;
            }
            currentToolCall = null;
          } else {
            // Text block stop — pass through
            yield event;
          }
          break;
        }

        default:
          // message_start, message_delta, message_stop, ping — pass through
          yield event;
          break;
      }
    }
  };

  return new Proxy(stream as object, {
    get(target, prop, receiver) {
      if (prop === Symbol.asyncIterator) {
        return () => guardedIterator();
      }
      return Reflect.get(target, prop, receiver);
    },
  }) as AsyncIterable<unknown>;
}
